### ISSUE TITLE
[Proposal]: "file private" visibility

### ISSUE THREAD

**Issue by stephentoub** [dotnet/csharplang#5529]
# "file private" visibility

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Enable source generators to emit code not visible to the rest of the assembly.

## Motivation
[motivation]: #motivation

Source generators frequently need to emit helper types that are only intended to be used by other generated source.  But there's frequently no way to emit such types in a way that completely shields them from being referenced by user code elsewhere in the application.  The generator can make them EditorBrowsable false, it can name them something awkward to use, but there's no great way to guarantee that user code won't access them, and thus that changes to those implementation details in a new version of the generator won't break existing code.

We should create a mechanism to enable this.  One such mechanism proposed by @jaredpar is a new "file private" visibility.  It would be like "internal", except the C# compiler would ensure it was only accessed by other types defined in the same file.

## Detailed design
[design]: #detailed-design

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by HaloFour** [comment]
Is a new accessibility modifier necessary or could the compiler be "relaxed" so that `private` can be applied to top-level types?

**Comment by HaloFour** [comment]
> What if you're generating part of a partial type?

Ah, I see, so this could also apply to a nested class emitted within the generated portion of a partial class, where `private` would not hide the nested class.

**Comment by jaredpar** [comment]
The idea of having file private types / methods is a mechanism I use a lot when coding in F#. There you can use a signature / implementation file pairing. Only types / methods that are present in the signature file are available to the rest of the code in the assembly. The rest are only visible to the code within the implementation file. 

I've found this to be very advantageous when coding in F# and I make use of it often. It's a way of defining helper types that just don't have any visibility within the rest of the assembly. It's a clean separation and really allows you to avoid spaghetti code in larger apps and instead force communication across defined boundaries. 

I've often longed for a similar feature in C#. When @stephentoub was chatting with me about the problems they are hitting isolating their generator code it occurred to me that this would also be a mechanism for generators to effectively isolate themselves. It gives them a greater degree of freedom in what they generate without the fear that consumers will depend on implementation details that generators never intended to expose.

**Comment by vladd** [comment]
Split into files has traditionally (except file-scoped namespaces) been neglectable in C#. Wouldn't it be better to make the types not file-private, but partial class part private? Other usecases seem to be covered by private inner classes.

**Comment by svick** [comment]
@vladd I don't think that would work well. Consider code like the following (using the `RegexGenerator` that's in the current .Net 7 alpha build):

```c#
partial class Server
{
    [RegexGenerator(@"(?:[0-9]{1,3}\.){3}[0-9]{1,3}", RegexOptions.IgnoreCase)]
    public partial Regex CreateIpAddressRegex(); 
}

partial class User
{
    [RegexGenerator("(.+)@(.+)")]
    public partial Regex CreateEmailRegex();
}
```

If the generated code for these two methods wanted to share some helper code, there is no way to use partial class part private types or private inner types. But file private types make it very easy.

**Comment by TahirAhmadov** [comment]
Is it possible to hide the generated code as nested private types inside a special class? I was going to say it's an easy _workaround_ but thinking about it more, it actually looks like a _solution_, if not **the** solution. This "container" class can be made partial, hence allowing the generation of a significantly large amount of "private" code without creating enormous files. In abstract terms (not `abstract` keyword), it makes sense, too - the inner workings of your generated functionality are hidden as private members of a "public" class (whether it's `public` or merely visible to the rest of the project).

**Comment by TahirAhmadov** [comment]
> @vladd I don't think that would work well. Consider code like the following (using the `RegexGenerator` that's in the current .Net 7 alpha build):
> 
> ```cs
> partial class Server
> {
>     [RegexGenerator(@"(?:[0-9]{1,3}\.){3}[0-9]{1,3}", RegexOptions.IgnoreCase)]
>     public partial Regex CreateIpAddressRegex(); 
> }
> 
> partial class User
> {
>     [RegexGenerator("(.+)@(.+)")]
>     public partial Regex CreateEmailRegex();
> }
> ```
> 
> If the generated code for these two methods wanted to share some helper code, there is no way to use partial class part private types or private inner types. But file private types make it very easy.

Good point. How about an attribute like `[PrivatesVisibleTo(typeof(User))]`?

**Comment by jnm2** [comment]
@TahirAhmadov That name, though

**Comment by CyrusNajmabadi** [comment]
I would be amenable to this idea. My personal take is that the symbols not be available outside the file *ever* (no attributes to expose it), and that they likely be implemented with a name mangling strategy to avoid cross file collisions.

**Comment by jaredpar** [comment]
The cross file symbol collision is definitely a consideration to take into account. It is a bit of a sore spot in other languages. IIRC F# gives a bit of a cryptic error when you have collisions instead of resolving them. 

I think we'd likely want to keep the simple names of the symbols the same. For example 

```c#
namespace Example { 
    file private class Widget { ... }
}
```

Think the identifier should still be `Widget` here. in metadata To do otherwise would end up subverting a lot of expectations. 

Instead think we should consider inserting a generated name between the namespace and the type name. In this case having `Widget` emitted as `Example.<>_generated.Widget`. That would allow us to avoid collisions while minimally subverting expectations around how types map to metadata names.

**Comment by CyrusNajmabadi** [comment]
@jaredpar .  I think that makes a lot of sense.  I agree with you that there is a strong desire to likely have the metadata name match (though it's not sacrosanct for me).  Your approach seems like a good best-of-both worlds option.  

For things like `field fields, if `<>_generated` was just a mutable struct, then this becomes effectively free at runtime (presuming the runtime doesn't fall off a cliff in any cases).  Are there things that would not work if these fields moved to such a struct?  Things like `ref` would still work and whatnot.  I presume FieldOffset might make no sense though... 

Just trying to think through corner cases here.  Thoughts?

**Comment by jaredpar** [comment]
That is a good point about fields. In my head I'd been focusing on the cases where the only type itself was marked as `file private` I hadn't really thought about the cases where you got down to the individual members (self fail)

I think at that level of granularity it gets really hard to maintain both:

1. Don't let different `file private` conflict with each other 
2. Don't break the mental model of developers on how code is emitted to metadata

Imagine for example reflection. The moment we start name mangling individual members then it becomes really hard to think about how to use features like reflection to use them.

Thinking along the lines of corner cases, I'm still trying to decide if `partial` makes sense when one of the declarations is `file private`. If the goal is that one `file private` is completely hidden from another then I don't think you can allow `partial`. Because then you run into cases like this: 

```c#
// file1 .cs 
file private partial class Widget { 
  public overrides string ToString() => "";
}

// file2.cs
partial class Widget { 
   // Error: multiple overrides of ToString
  public overrides string ToString() => "something";
```

Part of my brain is screaming "only allow `file private` on type declarations" because that allows us to give the guarantees we want and provides a clean solution for tools like generators that want an isolated space for their work.

**Comment by RikkiGibson** [comment]
Can't we use a modreq for this? Something like `class FilePrivate { }`?

**Comment by jaredpar** [comment]
@RikkiGibson you can't put a `modreq` on a type, only on method signatures. That is why we have to tricks like `[Obsolete]` when emitting a `ref struct`. 

A `modreq` could be a solution for non-virtual method conflicts though. The basic pattern we could employ is generating types into the assembly in the pattern of `Microsoft.CodeAnalysis.<>FilePrivateN` where `N` equals the number of files which contain at least one `file private` modifier. Then every method in a file gets a `modreq(MS.CA.<>FilePrivateN` which makes it unique. 

That does pose a problem though: such methods could not be used as implicit implementations of `interface` members. The presence of the `modreq` would make the signatures not match. We could counter by emitting an explicit implementation thunk under the hood though but we're back to that could conflict with explicit implementations in other files. 

We'd still need solutions for `virtual` methods, `interface` hookups, fields, etc ... 

The more we talk about this, and it's fantastic feedback, the more I'm wondering if forcing this at the type level is the best solution. It solves the core problems here, allows us to maintain the idea of it being truly private (don't worry abuot name conflicts) and no other language I'm aware of allowed parts of types to be `file private` (it's all or nothing).

**Comment by BhaaLseN** [comment]
I'm not really sure if this is actually needed (at least in the proposed way)? My first thought for making helper types unlikely to be used by non-generator code would simply be a nested namespace.

```c#
namespace Whatever.YourRegular.CodeUses
{
  partial class TypeToExtend
  {
    partial void DoTheThing() => DontUseThis.Helper.DoTheSharedThing();
    // or, if you dislike putting the namespace in front:
    //using DontUseThis;
    //partial void DoTheThing() => Helper.DoTheSharedThing();
  }
  namespace DontUseThis
  {
    internal static class Helper
    {
      public static void DoTheSharedThing() => throw null;
    }
  }
}
```

This doesn't _prevent_ non-generator code from using it, but it is more obvious if anyone _does_ attempt to use it, and it doesn't show up by default (since it isn't in the same namespace and you'd have to be `using` it first). Plus, it works today without any changes.

I do wonder if I could make use of the proposed solution anywhere (outside of source generators). I can't really think of any _other_ situation where this is both a useful addition and something that can't be done otherwise.
How would those types fare in the face of reflection? I briefly started typing about one thing I do that discovers types that implement a certain interface (and shouldn't _really_ be used by anyone else), but dropped it from my reply because it seems a fairly specialized and niche thing.

I _could_ see a top-level `class` using the `private` modifier to trigger this though (which today is not allowed and raises `CS1527`).
But if we go from there, it somewhat also opens the flood-gates for "invisible" base classes. Consolidate common functions in a non-`public` (or `internal`) class, derive from those, and callers see it as if the class had no base class (basically inlining it; but I don't know if IL could do this in any way without duplicating the code as well). And thats something for another day (and a different proposal, if at all).

**Comment by stephentoub** [comment]
> This doesn't prevent non-generator code from using it

And that's the problem. Then we change the generator, and code breaks.

**Comment by HaloFour** [comment]
@vladd 

Namespaces don't really exist, they're just part of the class name so they can't have their own metadata like accessibility levels.  The same namespace can be used freely between different assemblies as well.  The individual types within the namespace would need to carry that metadata.

**Comment by vladd** [comment]
@HaloFour 
Well, this seems to be not a problem in C++, it looks like the anonymous namespace has a unique unspeakable name from the linker point of view.

**Comment by HaloFour** [comment]
@vladd 

Different languages, different problems :)

Implicitly treating a `file private` class as within it's own unspeakable namespace might be a way for the compiler to lower the class as something that can't be referenced externally.  I don't know if that works for nested classes, though.  The relationship is defined by an entry in the NestedClass metadata table so perhaps the name of the nested class could be completely unrelated to the enclosing class, but you'd end up with something not expressable in C# or even IL which both require lexical nesting.

**Comment by vladd** [comment]
@HaloFour 
Well, I basically asked if the C++ approach is a way to go: no explicit file private modifier but using an anonymous namespace for it. This way we would avoid thinking in files but will be thinking in namespaces instead.

**Comment by PathogenDavid** [comment]
@vladd 

I feel like anonymous namespaces are borderline trivia in C++. It's not really obvious what they do unless you already know whereas something like `file private class` is more self-evident.

-----------

As an aside, personally I've found it to be valuable for generated source to be digestible by humans for debugging/learning purposes. I worry that this feature as proposed could end up forcing generator authors to put their entire output into one file (potentially making it painful for humans to interact with.)

It has its own downsides, but I wonder if friend types (https://github.com/dotnet/csharplang/discussions/2073) would be a better solution to this issue. (As-is, this whole proposal is basically a less flexible special-case of them.)

**Comment by lsoft** [comment]
Friends, let me share (may be) unpopular opinion: please do not complicate C# for minor reasons.  For SG-produced code the [EditorBrowsable](https://github.com/dotnet/roslyn/issues/51269) approach is enough, we need only make a final design and implement it. Also, @vladd 's proposal about **unspeakable** namespaces is better because of its lesser impact on language (as I see it).

if we want to fulfill the need without modifying the language - probalby it better to that... we have a "preprocessor" directives (`#nullable enable`), project settings (`<Nullable>Enable</Nullable) and visual editor level (EditorBrowsable). why we need to modify the language? let's put `#file private` on the top of SG-produced file! so the language will be modified about 1 new preprocessor directive... 

There is a lot of choices, please do not make C# more complex for the minor reason like hiding SG code.

Thanks!
(sorry for poor English, I hope my point is clear)

**Comment by teo-tsirpanis** [comment]
Regarding metadata representation, we can represent `file private` members with the ECMA.335 `compilercontrolled` visibility.

Having read the spec, my understanding is that type members with this visibility must be referenced with a `TypeDefinition` metadata token, making them effectively `internal`, but _very_ `internal`; not even `InternalsVisibleTo` or its evil twin we don't talk about can expose them to another ~~assembly~~ module, and that many compiler-controlled members in a type can have _the same name_, solving concerns about naming collisions.

Some outstanding questions are:

1. What about `file private` _types_? Only members may be marked as compiler-controlled. The types could be marked with a special attribute (`NotReferencableAttribute` or something like this), and/or have their names made unspeakable.
2. How would the debugger handle many type members with the same name? I ran a demo and both Visual Studio and Rider show two identically named fields like this:
![image](https://user-images.githubusercontent.com/12659251/148616206-8c0ca616-5b94-44a4-b6e5-9b6d8b18d997.png)
I couldn't exhibit how to test the debugger's reaction when I hover over a field name that has a duplicate.
3. How would the reflection handle compiler-controlled types with the same name? I ran a demo that calls `GetField("_x", BindingFlags.NonPublic | BindingFlags.Instance)`, it threw an `AmbiguousMatchException`.

I don't think (2) and (3) are very important issues; how likely is to have such collisions? But the runtime wouldn't bother and nor would the compiler who would totally ignore compiler-controlled members in referenced assemblies.

**Comment by acaly** [comment]
There has been many other proposals, especially related to member visibility, that has been rejected due to

> Language proposals which prevent specific syntax from occurring can be achieved with a Roslyn analyzer. Proposals that only make existing syntax optionally illegal will be rejected by the language design committee to prevent increased language complexity.

Can anyone here clearly explain why this proposal is not rejected for the same reason? We have been forced to use dirty attributes everywhere to disallow specific usage, and many of there are related with source generators. I don't think hiding something is anything different.

**Comment by HaloFour** [comment]
@acaly 

Because language team members are interested in this due to the use cases and scenarios that impact their work.  Language design is subjective and just because some proposals are rejected or accepted doesn't mean that all proposals of the same theme will meet that same fate.

**Comment by acaly** [comment]
Honestly I like this feature, because it can solve many scenarios that two or more closely related classes need to access internals of each other, while preventing those to be accessed by other classes in the same assembly.

The point is, this same problem has been raised by community long time ago, and is never seriously discussed simply because of the above reason. This proposal today proves that the claim

> Language proposals which prevent specific syntax from occurring can be achieved with a Roslyn analyzer.

is not always true, and you should really be more forgiving to proposals suggested by comminity, because the community has far more different use cases that you can't easily imagine right now but you may probably hit the same thing tomorrow.

**Comment by lsoft** [comment]
>  Language design is subjective and just because some proposals are rejected or accepted doesn't mean that all proposals of the same theme will meet that same fate.

Looks like there are 2 rule sets: the first for community (published), the second for members of Design Committee (unpusblished, a relaxed version of the first probably). Any proposals of specific kind given by community will be rejected, but same kind proposals from the members - the different situation. Interesting example of Orwell's doublethink, I guess.

I'm completely fine with that, honestly, I just like to know what the real rules are.

Because of the rules for community and design committee are different, there are a little value of spending time for us here.

**Comment by 333fred** [comment]
I want to clarify a couple of points here about "rules for thee but not for me":

1. Note that this proposal does _not_ have a champion. No one from the LDT has yet said they'd be willing to take it to the design meeting (though I am somewhat surprised @stephentoub has not done this, since it's his proposal).
2. Even if a proposal is championed, that does not mean that the LDT has said yay or nay to the feature. Different members of the LDT have different priorities and different goals for the language, and we try to build consensus about proposals. Just because a proposal has been made, and then a champion for that proposal found, does not mean that it will be implemented.
3. What LDT members are most interested in can change over time. Yes, this proposal has been made in the past, but the language has changed since that point (source generators, in particular, are a massive shift to the ecosystem). Remember that if time had been devoted to a "file private" feature years ago, then some other feature would have been cut, and we'd be talking about a completely different version of C#, which would have different idiomatic use patterns and a different set of priorities.

FWIW, I personally feel that this really just needs an IDE feature to update `EditorBrowsable` with a new flag, and perhaps an analyzer to actually error when APIs are called from non-source-generated code. I do not see this needing to be a language feature, with all the baggage that entails. So yes, the `Language proposals which prevent specific syntax from occurring can be achieved with a Roslyn analyzer.` is nearly always true, and is true here for me too.

**Comment by jaredpar** [comment]
> Language proposals which prevent specific syntax from occurring can be achieved with a Roslyn analyzer.

I've always interpreted this as requests along the lines of disabling `var`, `dynamic`, `await`, etc ... Basically literally a request for the ability to disable specific syntax in a given compilation. That is definitely the place for an analyzer because it's declaring that a given feature, and the runtime implications that come with it, are not desired in a given compilation. 

This is a bit of a different request. It's not an attempt to disable a specific type of syntax but rather restrict the visibility of parts of the compilation from other parts of the compilation. It's not focused on a particular piece of syntax but rather it's focused on the semantic aspect: which symbols are visible to other symbols. 

Yes this could be achieved with an analyzer, as could pretty much any request that asks that we restrict capabilities in the language. There are a few reasons why I think it's worth raising to the level of language support though:

1. This general rule is meant for scenarios where customers are looking to restrict a particular syntax in **their** compilations. Effectively it's a response to "we don't want X in our code base" and a reasonable response is "write an analyzer that restricts X". In this case though the analyzer isn't running in **their** code base, it's running in **my** code base. That is a bit different because ...
1. This is an analyzer that would likely be desired by many source generators. This is a request we've gotten several times through the design and impl of generators. Once it's know that there is an analyzer for it, it's reasonable to expect that many generator authors will incorporate it. Not doing it in the language means we end up in the real place where many generators are running this style of analyzer in the same compilation.
1. This is a feature that exists in many other languages. I've mentioned already that you can achieve this with Golang, F#, C, C++, etc ... This feature provides significant benefit to those platforms and it would provide the same to C#. You can see existing evidence that customers have wanted this over the years. Best example is the WPF editor in VS which uses a multi-build process to guarantee the exact type of isolation "file private" would provide. Roslyn has unit tests that provide similar enforcement parts of the assembly being invisible to others. 

3 is the most meaningful to me. The prior art in other languages and benefits it provide are compelling.

**Comment by 333fred** [comment]
I've published the notes from today's LDM: https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-14.md. In particular, with regards to the analyzer discussion, it's important to note that this proposal _does_ provide new expressiveness that C# doesn't allow today. For example:

```cs
// file1.cs
file private C {}
// file2.cs
file private C {}
```

With this new modifier, those would be unrelated types. In C# today, you couldn't duplicate that name and would have a compilation error.

**Comment by FaustVX** [comment]
@333fred 
But what happen if I also have `public class C { }` in my user code ?
Do I have an error for name conflict with a type that I can neither see, nor interact with ?

**Comment by 333fred** [comment]
> But what happen if I also have public class C { } in my user code ?

Unsure whether there would be a conflict there. That would still need to be designed.

**Comment by acaly** [comment]
> We could simply introduce `file private` as a new accessibility, or we could look at other combinations like `private internal`.

IMO, `private internal` is confusing, especially after we already have `private protected` and `internal protected`. It seems that you can couple any two of them and get a new visibility level, and those visibilities are quite unrelated.

**Comment by YegorStepanov** [comment]
Inner private classes can be moved out of the class. Usually they are located at the top of the class and need to be scrolled each time.

But `file private` is two-word declaration, and it's inconvenient for small helper classes (e.g. implementing `IComparable`/`IEqualityComparer`)

Maybe it should be one-word access modifier, like `file`?

**Comment by theunrepentantgeek** [comment]
A two-word declaration that's immediately apparent when would be far preferable to a single cryptic word that requires someone to refer to the definition. 

Optimizing keystrokes for the author of a file is the wrong way to go. We should instead optimize for immediate clarity when another developer reads the code.

> Usually [inner private classes] are located at the top of the class and need to be scrolled each time.

Not my experience - I don't think I've ever seen an inner private class at the top; As they're an internal implementation detail that's not visible to consumers, I've only ever seen them buried at the end of the class.

**Comment by RikkiGibson** [comment]
FWIW: I don't see *other languages are doing it* as innately convincing. In other contexts we answer that argument with: *but why is this right for C#?* If we have a satisfactory answer to that latter question, then certainly knowing *why* and *how* other languages have done the same feature can help us make better decisions in C#.

In my view the most compelling benefit of this feature is making it easier for source generators to introduce declarations that they *know* won't conflict with declarations in user code or from other generators. This is something you won't get from modifying `EditorBrowsable`, for example.

**Comment by CyrusNajmabadi** [comment]
>  but why is this right for C#?

I agree.  I only bring up other languages in comparison because i find this facility very nice for component design in other languages.  In C# i very much miss this because i often want a component to have an exposed area for other components to use, but then have impl pieces that should never be exposed.  My only choice here is to make thigns nested partials, and that's just not very ergonomic.

**Comment by RikkiGibson** [comment]
> I've published the notes from today's LDM: https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-14.md. In particular, with regards to the analyzer discussion, it's important to note that this proposal _does_ provide new expressiveness that C# doesn't allow today. For example:
> 
> ```cs
> // file1.cs
> file private C {}
> // file2.cs
> file private C {}
> ```
> 
> With this new modifier, those would be unrelated types. In C# today, you couldn't duplicate that name and would have a compilation error.

Is this only expected to provide "disambiguation" for type declarations? What happens in a scenario like the following:

```cs
// File1.cs
public partial class C
{
    file private void M() { System.Console.Write(1); }
}

// File2.cs
public partial class C
{
    file private void M() { System.Console.Write(2); }
}
```

Is this an error? Does name mangling of some kind occur during compilation?

**Comment by jaredpar** [comment]
> Is this an error? Does name mangling of non-type members occur during compilation?

Based on my take away this would be an error. The discussion was effectively that the `file` was simply a restriction on top of the other modifier. Effectively this is a `private` member that is further restricted that it can only be accessed directly from this file. The only level of name mangling discussed was for top level types.

Though it wasn't discussed explicitly I suspect most would be against mangling member names. The way in which type name mangling works is we don't actually mangle the short name, we simply mangle the namespace it's declared in. That means `file internal class C` is still emitted with the type name `C`, it's just the full name that is mangled. That was done in part to reduce surprises by keeping names the same. There isn't a similar mangling strategy that could be applied to member names.

This is all based on the idea though that `file <modifier>` is about restricting the accessibility of the type. There were those who wanted to think of it a different way, about making it removed / invisible. That would likely produce different answers here. The goal of the exercise we're doing at the moment is exploring the restricting accessibility route and what that would produce. 

This is an important scenario to call out though when we consider it.

**Comment by RikkiGibson** [comment]
IVTs present some challenges here.

```cs
// Assembly1
[assembly: InternalsVisibleTo("Assembly2")]

public class C1
{
    file internal static void M() { }
}

// Assembly2
public class C2
{
    void M()
    {
        C1.M(); // error
    }
}
```

Two ways I can see to make sure this error happens:
1. Mangle the name of `C1.M`.
2. Use a modreq on `C1.M` which hints that the method can't be used from outside the assembly.
    - I don't know if modreq can be used on fields. That could be problematic on `file internal` fields.
3. Use a well-known attribute or signature flag of some kind to hint that the member is "file internal" and thus from the runtime's POV it shouldn't be accessible outside the declaring assembly.

**Comment by jaredpar** [comment]
Couple other options to consider:

1. During metadata import the compiler can simply not import anything which was modified with `file`. That means the symbol simply isn't available. Or it can import it but still consider it `file internal` here and hence inaccessible. 
2. Could make the decision that members can't have the `file` modifier unless the containing type has it as well.

**Comment by RikkiGibson** [comment]
> 1. During metadata import the compiler can simply not import anything which was modified with `file`. That means the symbol simply isn't available. Or it can import it but still consider it `file internal` here and hence inaccessible.

Not sure how the compiler decides that something from metadata had the `file` modifier.

According to ECMA-335 II.23.2.4 FieldSig, fields can have custom modifiers. However, I couldn't find any compiler tests, etc. that used custom modifiers on fields.

https://github.com/stakx/ecma-335/blob/63957430afbca4bfb8b5df303f59aca9d978c35b/docs/ii.23.2.4-fieldsig.md

**Comment by jaredpar** [comment]
Could always use attributes.

**Comment by AraHaan** [comment]
I would rather have private types only visible to the same assembly, but ``InternalsVisibleTo`` would make it seem like the type never existed in terms of other projects referencing the assembly that defines it.

**Comment by ViIvanov** [comment]
As an idea to define "file-private" types I would suggest unnamed namespaces:
```cs
namespace Foo.Bar {
  internal partial class UserType  {
    internal partial void SomePartialMethod()  {
      My.Own.Stuff.Zoo.Blah(); // OK only in current file
      new Baz(); // OK only in current file
    }    
  }
}

namespace { // Unnamed namespace declaration, all internal types are "file-private"
  namespace My.Own.Stuff { // Named namespace inside unnamed, if needed
    class Zoo { // No access modifier, it is required. Or only optional "private" / "internal" allowed
      public static void Blah() { }
    }
  }

  class Baz { } // "Top-level" "file-private" enum
}
```

**Comment by AraHaan** [comment]
I got better idea for file private, and that is to allow ``private`` keyword on normal ClassDeclarationSyntax's that are a part of the current namespace and problem could be solved.

**Comment by CyrusNajmabadi** [comment]
@AraHaan that wouldn't solve a core case where you want the type hidden from code in the same assembly.  A private class would be visible to siblings (just as a private nested type is visible to it's siblings).

**Comment by jnm2** [comment]
@ViIvanov I like that, but it doesn't provide a solution for if you want only a nested type to be file-private.

**Comment by ViIvanov** [comment]
> @ViIvanov I like that, but it doesn't provide a solution for if you want only a nested type to be file-private.

Thank you. Sorry, I've missed this case.

**Comment by Pankraty** [comment]
Will this make it possible to create private extension methods? I was in a situation a couple of times when something might be nicely expressed as an extension method, but that would be too specific for exposing as a method visible for the entire assembly. Top-level private class would solve that problem.

**Comment by RikkiGibson** [comment]
I would expect to be able to define extension methods in a 'file' class and only be able to use them in the same file. That's a great scenario to test, thanks.

**Comment by BlinD-HuNTeR** [comment]
I don't wanna be the party pooper, but, what's wrong with nested types? If you are writing a source generator, and you don't want the generated types to be visible to the main assembly, just make all the generated sources be part of a big partial class, and the generated types as nested private types in that class.

In other words, that big class is just acting as a namespace for all purposes. Even source compatibility is preserved, you're just transforming a namespace into a class that spans multiple files. The fully-qualified names of all the nested types will remain the same.

This should definitely be mentioned in the "Alternatives" section of the proposal, after all, it's a "Do Nothing" alternative.

**Comment by CyrusNajmabadi** [comment]
@BlinD-HuNTeR 

"Do nothing" is always an implied alternative.  When we decide we want to do something, it's because we feel that the value has already proven itself over "do nothing" :)

**Comment by CyrusNajmabadi** [comment]
>In other words, that big class is just acting as a namespace for all purposes.

This came up as an option.  We felt that this wasn't ideal and that defining something just for this purpose is unpleasant.   We debated it but ended up here.

**Comment by stephentoub** [comment]
> I don't wanna be the party pooper, but, what's wrong with nested types? If you are writing a source generator, and you don't want the generated types to be visible to the main assembly, just make all the generated sources be part of a big partial class, and the generated types as nested private types in that class.

There's no way to hide it completely.  If the source generator is filling in a partial method _on the user's type_, that method needs access to the generated code; whatever it has access to, so will any other user code in type containing that partial method.

**Comment by eerhardt** [comment]
This was implemented in C# 11 right? https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/file.

Can this issue be closed?

**Comment by stephentoub** [comment]
@333fred, we keep these open until they're spec'd, yes?  Or is that tracked elsewhere?

**Comment by 333fred** [comment]
That's correct, but thank you for reminding me to label all the 11 issues appropriately.

**Comment by Xyncgas** [comment]
What if you are using a company git system and you are the one that can make change to your files and there is no easy way to clone your project because your git system is implemented in distributed performance with a blockchain backend therefore people are mad at you one day because you won't change your codes so that your methods are available to them which you are required by law to do so because it's the company's ip

**Comment by jrmoreno1** [comment]
@333fred: Could the check marks be added for prototype and implemented? I understand not closing until there is a documented spec, but I look for the check marks to know what stage it’s in

**Comment by 333fred** [comment]
Done.

**Comment by rob-ack** [comment]
Consider 'file' not only being useful for type definitions. In the scenario of "code generation" or/and "partial types" the file keyword is "useful on any member".
It would then allow to hide members which are needed for the generated or partial code to the other partial implementation details.

In my opinion file should be possible on any member as well (fields, properties, events, methods ...).

**Comment by acaly** [comment]
I found that issue that the IDE suggests me to convert some `[DllImport]` to `[LibraryImport]` on imported functions in a file scoped class. This is not possible because the generated implementation is in another source file, which has no access to the original definition.

**Comment by RikkiGibson** [comment]
> In my opinion file should be possible on any member as well (fields, properties, events, methods ...).

Absolutely agree. We just need to come to a consensus for how `file` relates to accessibility outside of top-level types.

I would personally like us to be able to "just see `file` as another access modifier", but I don't think the LDM necessarily agrees with me there. There is a temptation to start defining things like `file internal` (another type in same file can use it) and `file private`  (only the partial type declaration in the current file can use it), which starts to feel like a significant complexity burden.
